File: clients/typescript/src/utils/connection-string.ts
/**
 * Parse CortexDB connection strings
 * 
 * Format: cortexdb://[api_key@]host[:port]
 * 
 * Examples:
 * - cortexdb://localhost:8000
 * - cortexdb://my_key@localhost:8000
 * - cortexdb://my_key@api.cortexdb.com
 */

export interface ParsedConnection {
  baseUrl: string;
  apiKey?: string;
}

/**
 * Parse a CortexDB connection string
 * 
 * @param connectionString - Connection string in format cortexdb://[api_key@]host[:port]
 * @returns Parsed connection options
 * @throws Error if connection string is invalid
 * 
 * @example
 * ```typescript
 * parseConnectionString('cortexdb://localhost:8000')
 * // { baseUrl: 'http://localhost:8000' }
 * 
 * parseConnectionString('cortexdb://my_key@localhost:8000')
 * // { baseUrl: 'http://localhost:8000', apiKey: 'my_key' }
 * 
 * parseConnectionString('cortexdb://key@api.cortexdb.com:443')
 * // { baseUrl: 'https://api.cortexdb.com:443', apiKey: 'key' }
 * ```
 */
export function parseConnectionString(connectionString: string): ParsedConnection {
  // Remove cortexdb:// prefix
  if (!connectionString.startsWith('cortexdb://')) {
    throw new Error('Connection string must start with "cortexdb://"');
  }

  const withoutProtocol = connectionString.slice('cortexdb://'.length);
  
  // Split by @ to separate api_key from host
  let apiKey: string | undefined;
  let hostPart: string;

  if (withoutProtocol.includes('@')) {
    const parts = withoutProtocol.split('@');
    if (parts.length !== 2) {
      throw new Error('Invalid connection string format');
    }
    apiKey = parts[0];
    hostPart = parts[1];
  } else {
    hostPart = withoutProtocol;
  }

  // Parse host and port
  let host: string;
  let port: string | undefined;
  
  if (hostPart.includes(':')) {
    const hostPortParts = hostPart.split(':');
    host = hostPortParts[0];
    port = hostPortParts[1];
  } else {
    host = hostPart;
  }

  // Determine protocol (https for port 443 or production domains, http otherwise)
  const isSecure = port === '443' || 
                   host.includes('cortexdb.com') || 
                   (!host.includes('localhost') && !host.startsWith('127.'));
  
  const protocol = isSecure ? 'https' : 'http';
  
  // Build base URL
  let baseUrl = `${protocol}://${host}`;
  if (port && !(port === '443' && isSecure) && !(port === '80' && !isSecure)) {
    baseUrl += `:${port}`;
  }

  return {
    baseUrl,
    apiKey
  };
}

File: clients/python/cortexdb/connection_string.py
"""Parse CortexDB connection strings.

Format: cortexdb://[api_key@]host[:port]

Examples:
    - cortexdb://localhost:8000
    - cortexdb://my_key@localhost:8000
    - cortexdb://my_key@api.cortexdb.com
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class ParsedConnection:
    """Parsed connection string."""

    base_url: str
    api_key: Optional[str] = None


def parse_connection_string(connection_string: str) -> ParsedConnection:
    """Parse a CortexDB connection string.

    Args:
        connection_string: Connection string in format cortexdb://[api_key@]host[:port]

    Returns:
        Parsed connection options

    Raises:
        ValueError: If connection string is invalid

    Examples:
        >>> parse_connection_string('cortexdb://localhost:8000')
        ParsedConnection(base_url='http://localhost:8000', api_key=None)

        >>> parse_connection_string('cortexdb://my_key@localhost:8000')
        ParsedConnection(base_url='http://localhost:8000', api_key='my_key')

        >>> parse_connection_string('cortexdb://key@api.cortexdb.com:443')
        ParsedConnection(base_url='https://api.cortexdb.com:443', api_key='key')
    """
    # Remove cortexdb:// prefix
    if not connection_string.startswith("cortexdb://"):
        raise ValueError('Connection string must start with "cortexdb://"')

    without_protocol = connection_string[len("cortexdb://") :]

    # Split by @ to separate api_key from host
    api_key: Optional[str] = None
    host_part: str

    if "@" in without_protocol:
        parts = without_protocol.split("@")
        if len(parts) != 2:
            raise ValueError("Invalid connection string format")
        api_key = parts[0]
        host_part = parts[1]
    else:
        host_part = without_protocol

    # Parse host and port
    port: Optional[str] = None

    if ":" in host_part:
        host_port_parts = host_part.split(":")
        host = host_port_parts[0]
        port = host_port_parts[1]
    else:
        host = host_part

    # Determine protocol (https for port 443 or production domains, http otherwise)
    is_secure = (
        port == "443"
        or "cortexdb.com" in host
        or (not host.startswith("localhost") and not host.startswith("127."))
    )

    protocol = "https" if is_secure else "http"

    # Build base URL
    base_url = f"{protocol}://{host}"
    if port and not (port == "443" and is_secure) and not (port == "80" and not is_secure):
        base_url += f":{port}"

    return ParsedConnection(base_url=base_url, api_key=api_key)

